render-ws-spark-client/src/main/java/org/janelia/render/client/spark/intensityadjust/ShadingCorrectionTileClient.java

```diff
@@ -0,0 +1,245 @@
+            final Bounds bounds
+    ) throws IOException {
+
+        final RenderDataClient renderClient = parameters.webservice.getDataClient();
+        final ResolvedTileSpecCollection rtsc = renderClient.getResolvedTiles(parameters.stack, z);
+
+        if (layerModel == null) {
+            LOG.warn("No model found for z={}", z);
+        } else {
+            final Collection<TileSpec> tileSpecs = rtsc.getTileSpecs();
+            LOG.info("Adding shading correction for {} tile specs, z={}", tileSpecs.size(), z);
+
+            for (final TileSpec tileSpec : tileSpecs) {
+                addShadingCorrectionToTileSpec(tileSpec, layerModel, bounds);
+            }
+        }
+
+        renderClient.saveResolvedTiles(rtsc, parameters.targetStack, z);
+    }
+
+    private void addShadingCorrectionToTileSpec(
+        final TileSpec tileSpec,
+        final ShadingModel layerModel,
+        final Bounds bounds
+    ) {
+		// get uniform grid of points in tile
+        final int nSamples = (int) Math.ceil(Math.sqrt(layerModel.getMinNumMatches()));
+        final List<double[]> points = uniformGrid(nSamples);
+
+        final double centerX = bounds.getMinX() + 0.5 * bounds.getWidth();
+        final double centerY = bounds.getMinY() + 0.5 * bounds.getHeight();
+
+        final CoordinateTransformList<CoordinateTransform> transforms = tileSpec.getTransformList();
+        final List<PointMatch> matches = new ArrayList<>();
+
+        final double[] transformedPoint = new double[2];
+        for (final double[] tilePointNormalized : points) {
+            // transform normalized tile point (in [-1, 1] x [-1, 1]) to tile coordinates
+            transformedPoint[0] = (tilePointNormalized[0] + 1) * tileSpec.getWidth() / 2.0;
+            transformedPoint[1] = (tilePointNormalized[1] + 1) * tileSpec.getHeight() / 2.0;
+
+            // transform tile point to global coordinate system
+            transforms.applyInPlace(transformedPoint);
+
+            // transform global coordinate to [-1, 1] x [-1, 1] wrt to the layer bounds
+            transformedPoint[0] = (transformedPoint[0] - centerX) / (bounds.getWidth() / 2.0);
+            transformedPoint[1] = (transformedPoint[1] - centerY) / (bounds.getHeight() / 2.0);
+
+            // scale coordinates to [-1, 1] to evaluate the global model
+            layerModel.applyInPlace(transformedPoint);
+            final double[] value = new double[] {transformedPoint[0], 0};
+            matches.add(new PointMatch(new Point(tilePointNormalized), new Point(value)));
+        }
+
+        final ShadingModel tileModel = layerModel.copy();
+		try {
+			tileModel.fit(matches);
+		} catch (final NotEnoughDataPointsException | IllDefinedDataPointsException e) {
+			throw new RuntimeException(e);
+		}
+
+        // convert to filter and add to tile spec
+        final ShadingCorrectionFilter filter = new ShadingCorrectionFilter(tileModel);
+        tileSpec.setFilterSpec(FilterSpec.forFilter(filter));
```

Comment:
Do you need to check if the tile already has a filter spec and then add to a combined filter model here?

Reply:
This will be a common enough case that it might be best to create an `addFilterSpec` method to TileSpec that handles it.

Reply:
Yes, you are right. I vaguely remember writing such a thing in the past, but I can't find it.. Do you maybe remember in which context I might have done this already?

Reply:
I don't - sorry.  I just did a search.  Maybe [this code](https://github.com/saalfeldlab/render/blob/newsolver/render-ws-java-client/src/main/java/org/janelia/render/client/newsolver/DistributedIntensityCorrectionSolver.java#L315) is what you were remembering?

Reply:
Yes, that's it, thanks!